LoopKit/RemoteDataService.swift

```diff
@@ -7,30 +7,572 @@
 //
 
 import Foundation
+import HealthKit
+
+// MARK: - RemoteDataService
+
+public protocol RemoteDataServiceDelegate: AnyObject {
+
+    /// The delegate to perform status remote data queries.
+    var statusRemoteDataQueryDelegate: StatusRemoteDataQueryDelegate? { get }
+
+    /// The delegate to perform settings remote data queries.
+    var settingsRemoteDataQueryDelegate: SettingsRemoteDataQueryDelegate? { get }
+
+    /// The delegate to perform glucose remote data queries.
+    var glucoseRemoteDataQueryDelegate: GlucoseRemoteDataQueryDelegate? { get }
+
+    /// The delegate to perform dose remote data queries.
+    var doseRemoteDataQueryDelegate: DoseRemoteDataQueryDelegate? { get }
+
+    /// The delegate to perform carb remote data queries.
+    var carbRemoteDataQueryDelegate: CarbRemoteDataQueryDelegate? { get }
+
+}
 
 public protocol RemoteDataService: Service {
 
-    func uploadSettings(_ settings: Settings, lastUpdated: Date)
+    /// The delegate for the remote data service.
+    var remoteDataServiceDelegate: RemoteDataServiceDelegate? { get set }
+
+    /// Synchronize all remote data with the service.
+    func synchronizeRemoteData(completion: @escaping (_ result: Result<Bool, Error>) -> Void)
+
+}
+
+// MARK: - RemoteDataError
+
+/**
+ An error that can occur during a remote data operation.
+ */
+public enum RemoteDataError: Error {
+
+    /// Any query failure.
+    case queryFailure(Error)
+
+    /// A query timeout.
+    case queryTimeout
+
+    /// Any network failure.
+    case networkFailure(Error)
+
+    /// A network timeout.
+    case networkTimeout
+
+}
+
+// MARK: - StoredAndDeletedRemoteData
+
+/**
+ A remote data query can return data that must be stored and data that must be deleted. This
+ struct captures both into a single entity. The type of the stored and deleted data array is
+ generic.
+ */
+public struct StoredAndDeletedRemoteData<S, D> {
+
+    public var stored: [S]
+
+    public var deleted: [D]
+
+    public init() {
+        self.stored = []
+        self.deleted = []
+    }
+
+    public var isEmpty: Bool {
+        return stored.isEmpty && deleted.isEmpty
+    }
+
+    public var count: Int {
+        return stored.count + deleted.count
+    }
+
+}
+
+// MARK: - QueryAnchor
+
+/**
+ A query anchor must be explicitly initializable and raw representable.
+ */
+public protocol QueryAnchor: RawRepresentable {
+
+    init()
+
+}
+
+// MARK: - CoreDataQueryAnchor
+
+/**
+ A basic core data query anchor that uses a modification counter to track changes.
+ */
+public struct CoreDataQueryAnchor: QueryAnchor {
```

Comment:
I don't think this belongs in RemoteDataService. It should be part of the API for the data stores.

Reply:
Done.